### PR TITLE
Fix back and forth column dnd

### LINKED ISSUE #1567: Dragging a card back-and-forth between a column and it's neighbor acts weird
## Issue - Bug

https://parabol.slack.com/files/U0VV6QW64/F8F01JTFU/kazam_screencast_00003.mp4

- **Estimated effort:** 12 points ([see CONTRIBUTING.md](https://github.com/ParabolInc/action/blob/master/CONTRIBUTING.md#points-and-sizes))

### PR DESCRIPTION
**Pull request by dan-f**
This fixes #1567.

To verify:
- Set up your columns as follows:
```
| [ c ] | [ a ] |
|       | [ b ] |
```
- all in one drag, pull card `a` on top of `c`, and then back to the right where it used to be; where `b` now is.
- the preview card should follow the drag as expected.
- all other dragging and dropping should work as expected!

### CODE REVIEW

**Comment by mattkrick** [comment]
I gotta hand it to ya, your DnD setup is pretty sweet! i love the flex container at the end, i love `previousDraggedProjectId` that obviates the need for a throttle, I love that you got rid of my hacky `dragState`! So, seeing a similar `dragCache` crop up again gave me the sads.

I fixed it by referencing the updated project instead of the often-outdated object kept in `monitor.getItem()` for more details, see #1586. Feel free to borrow or base from it. You greatly simplified the DnD logic & now I can't bear to go back to keeping magic state. 

PS All the other refactoring on this looks good!

**Comment by dan-f** [comment]
Hey @mattkrick thanks for taking a look! :pray:

To be clear - the `DragCache` is simply a more general version of the `previousDraggedProjectId`; it's just a piece of memory used as state for memoizing our hover handlers. In other words, variables like `previousDraggedProjectId` vs `DragCache` are both "caches" in the sense that they're heap memory for storing answers to old questions. Since we replaced time-based mutation throttling in favor of a memoization-based solution, we'll have to keep track of state somewhere, whether we're using ad-hoc variables or a `Map`. This code change lifts the memoization strategy into a reusable pattern.

This contrasts to the `DragState` class, which was used for passing around knowledge of painted element coordinates as well as react component instances.

As far as resolving #1567 goes, I'd say that if we anticipate adding more drag-and-drop UI with hover event handlers that need to be fast, then let's go with this solution. But if you still have a strong aversion to `DragCache`, I don't mind using your solution, since both seem to work!

File: src/universal/modules/teamDashboard/components/ProjectColumn/ProjectColumn.js [hunk 1]
```diff
@@ -140,10 +140,6 @@ class ProjectColumn extends Component {
     // the front or back of the list, this will be `undefined`.
     const boundingProject = projects[targetIndex + (before ? -1 : 1)];
     const sortOrder = sortOrderBetween(targetProject, boundingProject, draggedProject, before);
-    const noActionNeeded = sortOrder === draggedProject.sortOrder && draggedProject.status === status;
-    if (noActionNeeded) {
-      return;
-    }
     const updatedProject = {id: draggedProject.id, sortOrder, status};
```
Comment (mattkrick) [inline src/universal/modules/teamDashboard/components/ProjectColumn/ProjectColumn.js]: As `updateProject` is currently written, the `updatedAt` field isn't changed when just the sortOrder changes. This was from user feedback that dragging things around shouldn't count as an "update". However, if you include `status`, it _will_ count as an update.

**Comment by mattkrick** [comment]
I hear ya! I still don't like storing additional state on each DraggableProject if we don't need to, but it's not a showstopper.
I cleaned up #1586 if you'd like to compare implementations. It handles all the update logic in the same function & uses data from relay instead of storing its own, but alas performance is still indistinguishable between the 2 at ~50fps on chrome incognito. 

If you do stick with this one, I ask that you only send the server `status` when `status` changes, otherwise it'll affect the `updatedAt` field, which isn't what we want for in-column drags (see comment).

**Pull request merged**